File: Assets/Plugins/Android/src/com/playphone/multinet/unity/MNDirectUnity.java
package com.playphone.multinet.unity;

import com.playphone.multinet.*;
import com.unity3d.player.UnityPlayer;
import com.playphone.multinet.core.MNSession;

public class MNDirectUnity {
    public static void init(final int param1_int, final String param2_String) {
        MNUnity.MARK();

        MNUnity.DLog("UnityPlayer.currentActivity class: " + UnityPlayer.currentActivity.getClass().getName());
        
        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.init(param1_int, param2_String,new MNDirectUnity.MNDirectEventHandler(),UnityPlayer.currentActivity);
                MNDirect.handleApplicationIntent(UnityPlayer.currentActivity.getIntent());
            }
        });
    }

    public static void shutdownSession() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.shutdownSession();
            }
        });
    }

    public static boolean isOnline() {
        MNUnity.MARK();

        return MNDirect.isOnline();
    }

    public static boolean isUserLoggedIn() {
        MNUnity.MARK();

        return MNDirect.isUserLoggedIn();
    }

    public static int getSessionStatus() {
        MNUnity.MARK();

        return MNDirect.getSessionStatus();
    }

    public static void postGameScore(final long score) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.postGameScore(score);
            }
        });
    }

    public static void postGameScorePending(final long score) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.postGameScorePending(score);
            }
        });
    }

    public static void cancelGame() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.cancelGame();
            }
        });
    }

    public static void setDefaultGameSetId(final int gameSetId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.setDefaultGameSetId(gameSetId);
            }
        });
    }

    public static int getDefaultGameSetId() {
        MNUnity.MARK();

        return MNDirect.getDefaultGameSetId();
    }

    public static void sendAppBeacon(final String actionName, final String beaconData) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.sendAppBeacon(actionName, beaconData);
            }
        });
    }

    public static void execAppCommand(final String name, final String param) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.execAppCommand(name, param);
            }
        });
    }

    public static void sendGameMessage(final String message) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.sendGameMessage(message);
            }
        });
    }

    protected static class MNDirectEventHandler extends com.playphone.multinet.MNDirectEventHandlerAbstract {

        @Override
        public void mnDirectDoStartGameWithParams(final MNGameParams params) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectDoStartGameWithParams",params);
                }
            });
        }

        @Override
        public void mnDirectDoFinishGame() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectDoFinishGame");
                }
            });
        }

        @Override
        public void mnDirectDoCancelGame() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectDoCancelGame");
                }
            });
        }

        @Override
        public void mnDirectViewDoGoBack() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectViewDoGoBack");
                }
            });
        }

        @Override
        public void mnDirectDidReceiveGameMessage(final String message, final MNUserInfo sender) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectDidReceiveGameMessage",message, sender);
                }
            });
        }

        @Override
        public void mnDirectSessionStatusChanged(final int newStatus) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectSessionStatusChanged",newStatus);
                }
            });
        }

        @Override
        public void mnDirectErrorOccurred(final MNErrorInfo error) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectErrorOccurred",error);
                }
            });
        }

        @Override
        public void mnDirectSessionReady(MNSession session) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_mnDirectSessionReady");
                }
            });
        }
    }
}

File: Assets/Plugins/iOS/PlayPhoneMultiNet/unity/MNDirectUnity.m
//
//  MNDirectUnity.m
//  PlayPhone SDK for Unity
//
//  Copyright (c) 2012 PlayPhone Inc. All rights reserved.
//

#import "MNDirect.h"

#import "MNDirectUnity.h"

@interface MNDirectUnity : NSObject <MNDirectDelegate>
+ (id)shared;
- (void)mnDirectDoStartGameWithParams:(MNGameParams*)params;
- (void)mnDirectDoFinishGame;
- (void)mnDirectDoCancelGame;
- (void)mnDirectViewDoGoBack;
- (void)mnDirectDidReceiveGameMessage:(NSString*)message from:(MNUserInfo*)sender;
- (void)mnDirectSessionStatusChangedTo:(int)newStatus;
- (void)mnDirectErrorOccurred:(MNErrorInfo*)error;
- (void)mnDirectSessionReady:(MNSession*) session;

@end

static MNDirectUnity *MNDirectUnityInstance = nil;

@implementation MNDirectUnity

+(id)shared {
    if (MNDirectUnityInstance == nil) {
        MNDirectUnityInstance = [[MNDirectUnity alloc]init];
    }
    
    return MNDirectUnityInstance;
}

- (void)mnDirectDoStartGameWithParams:(MNGameParams*)params {
    [MNUnity callUnityFunction:@"MNUM_mnDirectDoStartGameWithParams" withParams:(params == nil ? [NSNull null] : params), nil];
}
- (void)mnDirectDoFinishGame {
    [MNUnity callUnityFunction:@"MNUM_mnDirectDoFinishGame" withParams:nil];
}
- (void)mnDirectDoCancelGame {
    [MNUnity callUnityFunction:@"MNUM_mnDirectDoCancelGame" withParams:nil];
}
- (void)mnDirectViewDoGoBack {
    [MNUnity callUnityFunction:@"MNUM_mnDirectViewDoGoBack" withParams:nil];
}
- (void)mnDirectDidReceiveGameMessage:(NSString*)message from:(MNUserInfo*)sender {
    [MNUnity callUnityFunction:@"MNUM_mnDirectDidReceiveGameMessage" withParams:(message == nil ? [NSNull null] : message), (sender == nil ? [NSNull null] : sender), nil];
}
- (void)mnDirectSessionStatusChangedTo:(int)newStatus {
    [MNUnity callUnityFunction:@"MNUM_mnDirectSessionStatusChanged" withParams:[NSNumber numberWithInt:newStatus], nil];
}
- (void)mnDirectErrorOccurred:(MNErrorInfo*)error {
    [MNUnity callUnityFunction:@"MNUM_mnDirectErrorOccurred" withParams:(error == nil ? [NSNull null] : error), nil];
}
- (void)mnDirectSessionReady:(MNSession*) session {
    [MNUnity callUnityFunction:@"MNUM_mnDirectSessionReady" withParams:nil];
}

@end

EXTERN_C
void _MNDirect_Init(int gameId, const char* gameSecret) {
  MARK;

  [MNDirect prepareSessionWithGameId:gameId gameSecret:NSStringWithUTFStringSafe(gameSecret) andDelegate:[MNDirectUnity shared]];
}

EXTERN_C
void _MNDirect_ShutdownSession() {
    MARK;

    [MNDirect shutdownSession];
}

EXTERN_C
bool _MNDirect_IsOnline() {
    MARK;

    return [MNDirect isOnline];
}

EXTERN_C
bool _MNDirect_IsUserLoggedIn() {
    MARK;

    return [MNDirect isUserLoggedIn];
}

EXTERN_C
int _MNDirect_GetSessionStatus() {
    MARK;

    return [MNDirect getSessionStatus];
}

EXTERN_C
void _MNDirect_PostGameScore(long long score) {
    MARK;

    [MNDirect postGameScore:score];
}

EXTERN_C
void _MNDirect_PostGameScorePending(long long score) {
    MARK;

    [MNDirect postGameScorePending:score];
}

EXTERN_C
void _MNDirect_CancelGame() {
    MARK;

    [MNDirect cancelGame];
}

EXTERN_C
void _MNDirect_SetDefaultGameSetId(int gameSetId) {
    MARK;

    [MNDirect setDefaultGameSetId:gameSetId];
}

EXTERN_C
int _MNDirect_GetDefaultGameSetId() {
    MARK;

    return [MNDirect getDefaultGameSetId];
}

EXTERN_C
void _MNDirect_SendAppBeacon(const char* actionName, const char* beaconData) {
    MARK;

    [MNDirect sendAppBeacon:NSStringWithUTFStringSafe(actionName) beaconData:NSStringWithUTFStringSafe(beaconData)];
}

EXTERN_C
void _MNDirect_ExecAppCommand(const char* name, const char* param) {
    MARK;

    [MNDirect execAppCommand:NSStringWithUTFStringSafe(name) withParam:NSStringWithUTFStringSafe(param)];
}

EXTERN_C
void _MNDirect_SendGameMessage(const char* message) {
    MARK;

    [MNDirect sendGameMessage:NSStringWithUTFStringSafe(message)];
}


